src/common/mem.c

```diff
@@ -1573,4 +1573,362 @@ umem_tx_publish(struct umem_instance *umm, struct umem_rsrvd_act *rsrvd_act)
+
+	if (page->pg_copying)
+		return -DER_CHKPT_BUSY;
+
+	if (page != end_page) {
+		/** Eventually, we can just assert equal here.  But until we have a guarantee that
+		 * no allocation will span a page boundary, we have to handle this case.  We should
+		 * never have to span multiple pages though.
+		 */
+		if (end_page->pg_copying)
+			return -DER_CHKPT_BUSY;
+		D_ASSERT((page + 1) == end_page);
+		start_addr = end_addr & ~UMEM_CACHE_PAGE_SZ_MASK;
+
+		touch_page(cache, end_page, wr_tx, start_addr, end_addr);
+		end_addr = start_addr - 1;
+	}
+
+	touch_page(cache, page, wr_tx, addr, end_addr);
+
+	return 0;
+}
+
+/** We can look into other methods later. For now, only handle a few in-flight pages
+ *  being checkpointed at a time.  This will limit how much memory we use in each
+ *  xstream for checkpointing.
+ */
+#define MAX_INFLIGHT 8
+struct umem_checkpoint_data {
+	d_list_t                 cd_link;
+	struct umem_store_iod    cd_store_iod;
+	d_sg_list_t              cd_sg_list;
+	/** Each page can have at most every other chunk so reserve enough space up front to
+	 *  handle that.
+	 */
+	struct umem_store_region cd_regions[UMEM_CACHE_BMAP_SZ / 2];
+	d_iov_t                  cd_iovs[UMEM_CACHE_BMAP_SZ / 2];
+	/** Handle for the underlying I/O operations */
+	daos_handle_t            cd_fh;
+};
+
+static void
+page2chkpt(struct umem_page *page, struct umem_checkpoint_data *chkpt_data)
+{
+	uint64_t *bits = &page->pg_bmap[0];
+	struct umem_store_iod *store_iod = &chkpt_data->cd_store_iod;
+	d_sg_list_t           *sgl       = &chkpt_data->cd_sg_list;
+	uint64_t               bmap;
+	int       i;
+	int                    lead_bits;
+	uint64_t  offset    = (uint64_t)page->pg_id << UMEM_CACHE_PAGE_SZ_SHIFT;
+	uint64_t               map_offset;
+	uint8_t  *page_addr = page->pg_addr;
+	int       nr        = 0;
+	int                    count;
+	uint64_t               mask;
+	uint64_t               bit;
+
+	page->pg_chkpt_data = chkpt_data;
+	page->pg_dirty      = 0;
+
+	for (i = 0; i < UMEM_CACHE_BMAP_SZ; i++) {
+		if (bits[i] == 0)
+			goto next_bmap;
+
+		bmap = bits[i];
+		do {
+			lead_bits  = __builtin_clzll(bmap);
+			map_offset = lead_bits << UMEM_CACHE_CHUNK_SZ_SHIFT;
+			count      = 0;
+			mask       = 0;
+			for (;;) {
+				bit = 1ULL << (63 - lead_bits);
+				if ((bmap & bit) == 0)
+					break;
+				mask |= bit;
+				count++;
+			}
+
+			store_iod->io_regions[nr].sr_addr = offset + map_offset;
+			store_iod->io_regions[nr].sr_size = count << UMEM_CACHE_CHUNK_SZ_SHIFT;
+			sgl->sg_iovs[nr].iov_len          = sgl->sg_iovs[nr].iov_buf_len =
+			    count << UMEM_CACHE_CHUNK_SZ_SHIFT;
+			sgl->sg_iovs[nr].iov_buf = page_addr + map_offset;
+			nr++;
+
+			bmap &= ~mask;
+		} while (bmap != 0);
+
+next_bmap:
+		offset += UMEM_CACHE_CHUNK_SZ << 6;
+		page_addr += UMEM_CACHE_CHUNK_SZ << 6;
+	}
+	sgl->sg_nr_out = sgl->sg_nr = nr;
+	store_iod->io_nr            = nr;
+}
+
+int
+umem_cache_checkpoint(struct umem_cache *cache, uint64_t wal_tx)
```

Comment:
This 'wal_tx' is supposed to be the "checkpoint to ID" passed in by caller? If it's not useful, we can remove it from the parameters.

Reply:
Yeah, it should be the last ID committed but not necessarily to the WAL.   It slightly more efficient to pass it in than to walk through the pages and calculate it.